typeof discards previous non-null checks

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.2.0-dev.20181025

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
unknown typeof object null

**Code**

Run the following code via `tsc --noEmit --strict test.ts `
```ts
declare const x: unknown;
x !== null && typeof x === 'object' && 'field' in x;
```

**Expected behavior:**
It should pass the type check as the first condition ensures `x` is not null and the second one narrows the type to `object | null`. Together with the previous non-null check it should result in the `object` type.

**Actual behavior:**
```
test.ts:2:42 - error TS2531: Object is possibly 'null'.

2 x && typeof x === 'object' && 'field' in x;
                                           ~
```

A workaround is to switch the order of `typeof` and the `x` truthiness check:
```ts
declare const x: unknown;
typeof x === 'object' && x !== null && 'field' in x;
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> Note: you need to enable `strictNullChecks` manually!
https://www.typescriptlang.org/play/#src=declare%20const%20x%3A%20unknown%3B%0D%0Ax%20!%3D%3D%20null%20%26%26%20typeof%20x%20%3D%3D%3D%20'object'%20%26%26%20'field'%20in%20x%3B%0D%0A

**Related Issues:** <!-- Did you find other bugs that looked similar? --> #27180

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -753,6 +753,7 @@ namespace ts {
         const nonInferrableAnyType = createIntrinsicType(TypeFlags.Any, "any", ObjectFlags.ContainsWideningType);
         const intrinsicMarkerType = createIntrinsicType(TypeFlags.Any, "intrinsic");
         const unknownType = createIntrinsicType(TypeFlags.Unknown, "unknown");
+        const nonNullUnknownType = createIntrinsicType(TypeFlags.Unknown, "unknown");
         const undefinedType = createIntrinsicType(TypeFlags.Undefined, "undefined");
         const undefinedWideningType = strictNullChecks ? undefinedType : createIntrinsicType(TypeFlags.Undefined, "undefined", ObjectFlags.ContainsWideningType);
         const optionalType = createIntrinsicType(TypeFlags.Undefined, "undefined");
@@ -13995,7 +13996,9 @@ namespace ts {
             const includes = addTypesToUnion(typeSet, 0, types);
             if (unionReduction !== UnionReduction.None) {
                 if (includes & TypeFlags.AnyOrUnknown) {
-                    return includes & TypeFlags.Any ? includes & TypeFlags.IncludesWildcard ? wildcardType : anyType : unknownType;
+                    return includes & TypeFlags.Any ?
+                        includes & TypeFlags.IncludesWildcard ? wildcardType : anyType :
+                        includes & TypeFlags.Null || containsType(typeSet, unknownType) ? unknownType : nonNullUnknownType;
                 }
                 if (exactOptionalPropertyTypes && includes & TypeFlags.Undefined) {
                     const missingIndex = binarySearch(typeSet, missingType, getTypeId, compareValues);
@@ -22100,13 +22103,6 @@ namespace ts {
             return false;
         }
 
-        // Given a source x, check if target matches x or is an && operation with an operand that matches x.
-        function containsTruthyCheck(source: Node, target: Node): boolean {
-            return isMatchingReference(source, target) ||
-                (target.kind === SyntaxKind.BinaryExpression && (target as BinaryExpression).operatorToken.kind === SyntaxKind.AmpersandAmpersandToken &&
-                (containsTruthyCheck(source, (target as BinaryExpression).left) || containsTruthyCheck(source, (target as BinaryExpression).right)));
-        }
-
         function getPropertyAccess(expr: Expression) {
             if (isAccessExpression(expr)) {
                 return expr;
@@ -23119,7 +23115,8 @@ namespace ts {
             if (resultType === unreachableNeverType || reference.parent && reference.parent.kind === SyntaxKind.NonNullExpression && !(resultType.flags & TypeFlags.Never) && getTypeWithFacts(resultType, TypeFacts.NEUndefinedOrNull).flags & TypeFlags.Never) {
                 return declaredType;
             }
-            return resultType;
+            // The non-null unknown type should never escape control flow analysis.
+            return resultType === nonNullUnknownType ? unknownType : resultType;
 
             function getOrSetCacheKey() {
                 if (isKeySet) {
@@ -23607,7 +23604,8 @@ namespace ts {
 
             function narrowTypeByTruthiness(type: Type, expr: Expression, assumeTrue: boolean): Type {
                 if (isMatchingReference(reference, expr)) {
-                    return getTypeWithFacts(type, assumeTrue ? TypeFacts.Truthy : TypeFacts.Falsy);
+                    return type.flags & TypeFlags.Unknown && assumeTrue ? nonNullUnknownType :
+                        getTypeWithFacts(type, assumeTrue ? TypeFacts.Truthy : TypeFacts.Falsy);
                 }
                 if (strictNullChecks && assumeTrue && optionalChainContainsReference(expr, reference)) {
                     type = getTypeWithFacts(type, TypeFacts.NEUndefinedOrNull);
@@ -23765,7 +23763,7 @@ namespace ts {
                         valueType.flags & TypeFlags.Null ?
                             assumeTrue ? TypeFacts.EQNull : TypeFacts.NENull :
                             assumeTrue ? TypeFacts.EQUndefined : TypeFacts.NEUndefined;
-                    return getTypeWithFacts(type, facts);
+                    return type.flags & TypeFlags.Unknown && facts & (TypeFacts.NENull | TypeFacts.NEUndefinedOrNull) ? nonNullUnknownType : getTypeWithFacts(type, facts);
                 }
                 if (assumeTrue) {
                     const filterFn: (t: Type) => boolean = operator === SyntaxKind.EqualsEqualsToken ?
@@ -23795,15 +23793,10 @@ namespace ts {
                     return type;
                 }
                 if (assumeTrue && type.flags & TypeFlags.Unknown && literal.text === "object") {
-                    // The pattern x && typeof x === 'object', where x is of type unknown, narrows x to type object. We don't
-                    // need to check for the reverse typeof x === 'object' && x since that already narrows correctly.
-                    if (typeOfExpr.parent.parent.kind === SyntaxKind.BinaryExpression) {
-                        const expr = typeOfExpr.parent.parent as BinaryExpression;
-                        if (expr.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken && expr.right === typeOfExpr.parent && containsTruthyCheck(reference, expr.left)) {
-                            return nonPrimitiveType;
-                        }
-                    }
-                    return getUnionType([nonPrimitiveType, nullType]);
+                    // The non-null unknown type is used to track whether a previous narrowing operation has removed the null type
+                    // from the unknown type. For example, the expression `x && typeof x === 'object'` first narrows x to the non-null
+                    // unknown type, and then narrows that to the non-primitive type.
+                    return type === nonNullUnknownType ? nonPrimitiveType : getUnionType([nonPrimitiveType, nullType]);
                 }
                 const facts = assumeTrue ?
                     typeofEQFacts.get(literal.text) || TypeFacts.TypeofEQHostObject :

--- a/tests/baselines/reference/controlFlowTypeofObject.errors.txt
+++ b/tests/baselines/reference/controlFlowTypeofObject.errors.txt
@@ -1,0 +1,77 @@
+tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts(66,13): error TS2345: Argument of type 'object | null' is not assignable to parameter of type 'object'.
+  Type 'null' is not assignable to type 'object'.
+
+
+==== tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts (1 errors) ====
+    declare function obj(x: object): void;
+    
+    function f1(x: unknown) {
+        if (!x) {
+            return;
+        }
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    
+    function f2(x: unknown) {
+        if (x === null) {
+            return;
+        }
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    
+    function f3(x: unknown) {
+        if (x == null) {
+            return;
+        }
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    
+    function f4(x: unknown) {
+        if (x == undefined) {
+            return;
+        }
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    
+    function f5(x: unknown) {
+        if (!!true) {
+            if (!x) {
+                return;
+            }
+        }
+        else {
+            if (x === null) {
+                return;
+            }
+        }
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    
+    function f6(x: unknown) {
+        if (x === null) {
+            x;
+        }
+        else {
+            x;
+            if (typeof x === 'object') {
+                obj(x);
+            }
+        }
+        if (typeof x === 'object') {
+            obj(x);  // Error
+                ~
+!!! error TS2345: Argument of type 'object | null' is not assignable to parameter of type 'object'.
+!!! error TS2345:   Type 'null' is not assignable to type 'object'.
+        }
+    }
+    

--- a/tests/baselines/reference/controlFlowTypeofObject.js
+++ b/tests/baselines/reference/controlFlowTypeofObject.js
@@ -1,0 +1,144 @@
+//// [controlFlowTypeofObject.ts]
+declare function obj(x: object): void;
+
+function f1(x: unknown) {
+    if (!x) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f2(x: unknown) {
+    if (x === null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f3(x: unknown) {
+    if (x == null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f4(x: unknown) {
+    if (x == undefined) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f5(x: unknown) {
+    if (!!true) {
+        if (!x) {
+            return;
+        }
+    }
+    else {
+        if (x === null) {
+            return;
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f6(x: unknown) {
+    if (x === null) {
+        x;
+    }
+    else {
+        x;
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x);  // Error
+    }
+}
+
+
+//// [controlFlowTypeofObject.js]
+"use strict";
+function f1(x) {
+    if (!x) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+function f2(x) {
+    if (x === null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+function f3(x) {
+    if (x == null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+function f4(x) {
+    if (x == undefined) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+function f5(x) {
+    if (!!true) {
+        if (!x) {
+            return;
+        }
+    }
+    else {
+        if (x === null) {
+            return;
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+function f6(x) {
+    if (x === null) {
+        x;
+    }
+    else {
+        x;
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x); // Error
+    }
+}
+
+
+//// [controlFlowTypeofObject.d.ts]
+declare function obj(x: object): void;
+declare function f1(x: unknown): void;
+declare function f2(x: unknown): void;
+declare function f3(x: unknown): void;
+declare function f4(x: unknown): void;
+declare function f5(x: unknown): void;
+declare function f6(x: unknown): void;

--- a/tests/baselines/reference/controlFlowTypeofObject.symbols
+++ b/tests/baselines/reference/controlFlowTypeofObject.symbols
@@ -1,0 +1,136 @@
+=== tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts ===
+declare function obj(x: object): void;
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 0, 21))
+
+function f1(x: unknown) {
+>f1 : Symbol(f1, Decl(controlFlowTypeofObject.ts, 0, 38))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 2, 12))
+
+    if (!x) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 2, 12))
+
+        return;
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 2, 12))
+
+        obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 2, 12))
+    }
+}
+
+function f2(x: unknown) {
+>f2 : Symbol(f2, Decl(controlFlowTypeofObject.ts, 9, 1))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 11, 12))
+
+    if (x === null) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 11, 12))
+
+        return;
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 11, 12))
+
+        obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 11, 12))
+    }
+}
+
+function f3(x: unknown) {
+>f3 : Symbol(f3, Decl(controlFlowTypeofObject.ts, 18, 1))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 20, 12))
+
+    if (x == null) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 20, 12))
+
+        return;
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 20, 12))
+
+        obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 20, 12))
+    }
+}
+
+function f4(x: unknown) {
+>f4 : Symbol(f4, Decl(controlFlowTypeofObject.ts, 27, 1))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 29, 12))
+
+    if (x == undefined) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 29, 12))
+>undefined : Symbol(undefined)
+
+        return;
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 29, 12))
+
+        obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 29, 12))
+    }
+}
+
+function f5(x: unknown) {
+>f5 : Symbol(f5, Decl(controlFlowTypeofObject.ts, 36, 1))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 38, 12))
+
+    if (!!true) {
+        if (!x) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 38, 12))
+
+            return;
+        }
+    }
+    else {
+        if (x === null) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 38, 12))
+
+            return;
+        }
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 38, 12))
+
+        obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 38, 12))
+    }
+}
+
+function f6(x: unknown) {
+>f6 : Symbol(f6, Decl(controlFlowTypeofObject.ts, 52, 1))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+
+    if (x === null) {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+
+        x;
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+    }
+    else {
+        x;
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+
+        if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+
+            obj(x);
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+        }
+    }
+    if (typeof x === 'object') {
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+
+        obj(x);  // Error
+>obj : Symbol(obj, Decl(controlFlowTypeofObject.ts, 0, 0))
+>x : Symbol(x, Decl(controlFlowTypeofObject.ts, 54, 12))
+    }
+}
+

--- a/tests/baselines/reference/controlFlowTypeofObject.types
+++ b/tests/baselines/reference/controlFlowTypeofObject.types
@@ -1,0 +1,179 @@
+=== tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts ===
+declare function obj(x: object): void;
+>obj : (x: object) => void
+>x : object
+
+function f1(x: unknown) {
+>f1 : (x: unknown) => void
+>x : unknown
+
+    if (!x) {
+>!x : boolean
+>x : unknown
+
+        return;
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+    }
+}
+
+function f2(x: unknown) {
+>f2 : (x: unknown) => void
+>x : unknown
+
+    if (x === null) {
+>x === null : boolean
+>x : unknown
+>null : null
+
+        return;
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+    }
+}
+
+function f3(x: unknown) {
+>f3 : (x: unknown) => void
+>x : unknown
+
+    if (x == null) {
+>x == null : boolean
+>x : unknown
+>null : null
+
+        return;
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+    }
+}
+
+function f4(x: unknown) {
+>f4 : (x: unknown) => void
+>x : unknown
+
+    if (x == undefined) {
+>x == undefined : boolean
+>x : unknown
+>undefined : undefined
+
+        return;
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+    }
+}
+
+function f5(x: unknown) {
+>f5 : (x: unknown) => void
+>x : unknown
+
+    if (!!true) {
+>!!true : true
+>!true : false
+>true : true
+
+        if (!x) {
+>!x : boolean
+>x : unknown
+
+            return;
+        }
+    }
+    else {
+        if (x === null) {
+>x === null : boolean
+>x : unknown
+>null : null
+
+            return;
+        }
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+    }
+}
+
+function f6(x: unknown) {
+>f6 : (x: unknown) => void
+>x : unknown
+
+    if (x === null) {
+>x === null : boolean
+>x : unknown
+>null : null
+
+        x;
+>x : null
+    }
+    else {
+        x;
+>x : unknown
+
+        if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+            obj(x);
+>obj(x) : void
+>obj : (x: object) => void
+>x : object
+        }
+    }
+    if (typeof x === 'object') {
+>typeof x === 'object' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>'object' : "object"
+
+        obj(x);  // Error
+>obj(x) : void
+>obj : (x: object) => void
+>x : object | null
+    }
+}
+

--- a/tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts
+++ b/tests/cases/conformance/controlFlow/controlFlowTypeofObject.ts
@@ -1,0 +1,71 @@
+// @strict: true
+// @declaration: true
+
+declare function obj(x: object): void;
+
+function f1(x: unknown) {
+    if (!x) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f2(x: unknown) {
+    if (x === null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f3(x: unknown) {
+    if (x == null) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f4(x: unknown) {
+    if (x == undefined) {
+        return;
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f5(x: unknown) {
+    if (!!true) {
+        if (!x) {
+            return;
+        }
+    }
+    else {
+        if (x === null) {
+            return;
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x);
+    }
+}
+
+function f6(x: unknown) {
+    if (x === null) {
+        x;
+    }
+    else {
+        x;
+        if (typeof x === 'object') {
+            obj(x);
+        }
+    }
+    if (typeof x === 'object') {
+        obj(x);  // Error
+    }
+}

